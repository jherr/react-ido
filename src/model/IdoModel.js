import { observable, decorate } from 'mobx';
import uuid from 'uuid/v1';

export default class IdoModel {
  constructor(config, elements, onChange) {
    this.config = config;
    this.elements = elements;
    this.activeTarget = '';
    this.isDragging = false;
    this.editMode = {};
    this.onChange = onChange || (() => {});
  }

  toggleEditMode(id) {
    this.editMode[id] = !this.editMode[id];
  }

  onBeginDrag() {
    this.activeTarget = '';
    this.isDragging = true;
  }

  onEndDrag() {
    this.activeTarget = '';
    this.isDragging = false;
  }

  onDelete(id) {
    const {
      row,
      index
    } = this.findItem(id);
    this.elements[row].splice(index, 1);
    this.pruneElements();
    this.onChange(this.elements);
  }

  onDrop(source, target) {
    this.activeTarget = '';
    if (source.source === 'palette') {
      this.insertElement(target, this.newElement(source.widget));
    } else if(source.id) {
      const {
        row,
        index
      } = this.findItem(source.id);

      const elementData = Object.assign( this.elements[row][index] );
      this.elements[row].splice(index, 1);
      this.insertElement(target, elementData);
    }

    this.pruneElements();
    this.onChange(this.elements);
  }

  pruneElements() {
    const findEmptyRowIndex = () => {
      let found = null;
      this.elements.forEach((row, index) => {
        if (row.length === 0) {
          found = index;
        }
      });
      return found;
    }

    let emptyRow = findEmptyRowIndex();
    while(emptyRow !== null) {
      this.elements.splice(emptyRow, 1);
      emptyRow = findEmptyRowIndex();
    }
  }

  insertElement(target, element) {
    if (target.id === 'after') {
      this.elements.push([ element ]);
    } else if (target.beforeRow !== undefined) {
      this.elements.splice(
        target.beforeRow,
        0,
        [ element ]
      );
    } else if (target.beforeElement) {
      const {
        row,
        index
      } = this.findItem(target.beforeElement);
      this.elements[row].splice(index, 0, element);
    } else if (target.afterElement) {
      const {
        row,
        index
      } = this.findItem(target.afterElement);
      this.elements[row].splice(index + 1, 0, element);
    }
  } 

  findItem(id) {
    const out = {
      row: null,
      index: null
    };
    this.elements.forEach((rowElements, row) => {
      rowElements.forEach((element, index) => {
        if (element.id === id) {
          out.row = row;
          out.index = index;
        }
      });
    });
    return out;
  }

  newElement(widget) {
    return {
      type: widget.type,
      id: uuid(),
      data: Object.assign(widget.defaultValues),
      width: 12,
    };
  }
}

decorate(IdoModel, {
  config: observable,
  elements: observable,
  activeTarget: observable,
  isDragging: observable,
  editMode: observable,
});
