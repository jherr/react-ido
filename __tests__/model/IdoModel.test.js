import { toJS } from 'mobx';

import IdoModel from '../../src/model/IdoModel';

const fixElements = elements => toJS(elements).map(row =>
  row.map((elem) => {
    delete elem.id;
    return elem;
  })
);

describe('IdoModel', () => {
  it('should construct', () => {
    const model = new IdoModel({
      a: true,
    }, [], () => {});
    expect(model.config.a).toBe(true);
  });

  it('should handle togging edit mode', () => {
    const model = new IdoModel({}, [], () => {});
    expect(model.editMode.a).toBe(undefined);
    model.toggleEditMode('a');
    expect(model.editMode.a).toBe(true);
    model.toggleEditMode('a');
    expect(model.editMode.a).toBe(false);
  });

  it('should handle starting a drag', () => {
    const model = new IdoModel({}, [], () => {});
    model.onBeginDrag();
    expect(model.isDragging).toBe(true);
  });

  it('should handle ending a drag', () => {
    const model = new IdoModel({}, [], () => {});
    model.onEndDrag();
    expect(model.isDragging).toBe(false);
  });

  it('should allow a delete', () => {
    const model = new IdoModel({}, [
      [
        {
          id: 'a',
        },
      ],
    ], () => {});
    model.onDelete('a');
    expect(model.elements).toEqual([]);
  });

  it('should handle a palette drop at the end', () => {
    const model = new IdoModel({}, [
      [
      ],
    ], () => {});
    model.onDrop({
      source: 'palette',
      widget: {
        type: 'foo',
        defaultValues: {},
      },
    }, {
      id: 'after',
    });
    expect(fixElements(model.elements)).toMatchSnapshot();
  });

  it('should handle a palette drop at the front', () => {
    const model = new IdoModel({}, [
      [
      ],
    ], () => {});
    model.onDrop({
      source: 'palette',
      widget: {
        type: 'foo',
        defaultValues: {},
      },
    }, {
      beforeRow: 0,
    });
    expect(fixElements(model.elements)).toMatchSnapshot();
  });

  it('should handle a palette drop before an element', () => {
    const model = new IdoModel({}, [
      [
        {
          id: 'a',
        },
      ],
    ], () => {});
    model.onDrop({
      source: 'palette',
      widget: {
        type: 'foo',
        defaultValues: {},
      },
    }, {
      beforeElement: 'a',
    });
    expect(fixElements(model.elements)).toMatchSnapshot();
  });

  it('should handle a palette drop after an element and fire onChange', () => {
    const spy = jest.fn();
    const model = new IdoModel({}, [
      [
        {
          id: 'a',
        },
      ],
    ], spy);
    model.onDrop({
      source: 'palette',
      widget: {
        type: 'foo',
        defaultValues: {},
      },
    }, {
      afterElement: 'a',
    });
    expect(fixElements(model.elements)).toMatchSnapshot();
    expect(spy).toBeCalled();
  });

  it('should ignore drops it doesn\'t understand', () => {
    const model = new IdoModel({}, [
      [
        {
          id: 'a',
        },
      ],
    ], () => {});
    model.onDrop({
      source: 'palette',
      widget: {
        type: 'foo',
        defaultValues: {},
      },
    }, {
    });
    expect(fixElements(model.elements)).toMatchSnapshot();
  });

  it('should handle a palette drop after an element', () => {
    const model = new IdoModel({}, [
      [
        {
          id: 'a',
        },
      ],
      [
        {
          id: 'b',
        },
      ],
    ]);
    model.onDrop({
      id: 'a',
      widget: {
        type: 'foo',
        defaultValues: {},
      },
    }, {
      beforeElement: 'b',
    });
    expect(fixElements(model.elements)).toMatchSnapshot();
  });

  it('should ignore a palette drop it doesn\'t understand', () => {
    const model = new IdoModel({}, []);
    model.onDrop({}, {});
    expect(fixElements(model.elements)).toMatchSnapshot();
  });
});