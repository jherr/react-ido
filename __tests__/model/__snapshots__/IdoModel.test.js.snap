// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IdoModel should handle a palette drop after an element 1`] = `
Array [
  Array [
    Object {},
    Object {},
  ],
]
`;

exports[`IdoModel should handle a palette drop after an element and fire onChange 1`] = `
Array [
  Array [
    Object {},
    Object {
      "data": Object {},
      "type": "foo",
      "width": 12,
    },
  ],
]
`;

exports[`IdoModel should handle a palette drop at the end 1`] = `
Array [
  Array [
    Object {
      "data": Object {},
      "type": "foo",
      "width": 12,
    },
  ],
]
`;

exports[`IdoModel should handle a palette drop at the front 1`] = `
Array [
  Array [
    Object {
      "data": Object {},
      "type": "foo",
      "width": 12,
    },
  ],
]
`;

exports[`IdoModel should handle a palette drop before an element 1`] = `
Array [
  Array [
    Object {
      "data": Object {},
      "type": "foo",
      "width": 12,
    },
    Object {},
  ],
]
`;

exports[`IdoModel should ignore a palette drop it doesn't understand 1`] = `Array []`;

exports[`IdoModel should ignore drops it doesn't understand 1`] = `
Array [
  Array [
    Object {},
  ],
]
`;
